Add vignette on uploading data to AnVIL from a local computer (#17)
fhdsl/Data_on_AnVIL: 02-bringing_own_data_local_computer.Rmd

```{r, include = FALSE}
ottrpal::set_knitr_image_path()
knitr::opts_chunk$set(out.width = "100%")
```
# From Your Local Computer {#uploading-local-computer}

> When moving to cloud computing, you may have data that you've stored on a local server that you need to transfer into the cloud. AnVIL has developed a simple drag-and-drop tool called the Data Uploader to make this process quick and painless.

Reviewer: ```suggestion
When moving to cloud computing, you may have data that you've stored on a local computer (such as a work laptop) that you need to transfer into the cloud. AnVIL has developed a simple drag-and-drop tool called the Data Uploader to make this process quick and painless.
```
?

```

```{r, echo=FALSE, fig.alt='Image shows a screenshot of the workspace Data tab. The Files directory and link on the bottom left is highlighted. This image shows the collection called covid_membrane.'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3c7f7fdd14a_0_27#slide=id.g3c7f7fdd14a_0_27")
```

If you click on the name of the file, you can also see the details of the file.

```{r, echo=FALSE, fig.alt='Image shows a screenshot of the details of the files we successfully transferred.'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3c7f7fdd14a_0_30#slide=id.g3c7f7fdd14a_0_30")
```

This is a good time to check that the sizes of the files you transferred match the sizes of the original files.

## Summary

- Create a workspace
- Open Data Uploader
- Upload data and metadata (this requires creating a data collection)
- Verify that files have been transferred

## Additional Resources

Check out these AnVIL Portal articles for more details!

> - You can read more about the Data Uploader [here](https://support.terra.bio/hc/en-us/articles/4419428208411-Upload-data-and-populate-the-table-with-linked-file-paths) 

Reviewer: ```suggestion
- You can read more about the Data Uploader [here](https://support.terra.bio/hc/en-us/articles/4419428208411-Upload-data-and-populate-the-table-with-linked-file-paths). 
```

```{r, include = FALSE}
ottrpal::set_knitr_image_path()
knitr::opts_chunk$set(out.width = "100%")
```
# From Your Local Computer {#uploading-local-computer}

> When moving to cloud computing, you may have data that you've stored on a local computer (such as a work laptop) that you need to transfer into the cloud. AnVIL has developed a simple drag-and-drop tool called the Data Uploader to make this process quick and painless.

Reviewer 2: My first reaction when hearing drag-and-drop is to expect this means I'll be moving files manually one-by-one. If I had a lot of files, I'd immediately start wondering just how painful this is going to be for me.

A couple of thoughts:
1. Perhaps clarify that the Data Uploader can handle whole (flat?) directories - e.g. if you had a directory with 100+ fasta files, it would still be pretty easy
2. Perhaps link to the hpc chapter for people who would prefer to use the command line?

Author: Good point, have fixed!

> In this example, we'll upload some genomic data that is stored on your local computer. You can use a file of your own, or download the files we are using in this example from our GitHub repo.

Reviewer 2: ```suggestion
In this example, we'll upload some genomic data that is stored on your local computer. You can use a file of your own, or download the files we are using in this example from our [GitHub repo](https://github.com/fhdsl/Data_on_AnVIL/tree/main/data).
```

> The data files for this vignette are very small fasta files. Each file contains only a single gene from a single CoVID variant (initially downloaded onto a local server from GenBank). While we kept these files small for the vignette, the AnVIL Data Uploader can easily handle much larger files.

Reviewer 3: ```suggestion
The data files for this vignette are very small fasta files. Each file contains only a single gene from a single COVID variant (initially downloaded onto a local server from GenBank). While we kept these files small for the vignette, the AnVIL Data Uploader can easily handle much larger files.
```

::: {.notice}
_Genetics_

**Novice**: no genetics skills needed

_Programming skills_

**Novice**: no programming skills needed
:::

::: {.notice}
**What will this cost?**

In general, it is free to upload data to the cloud, but the workflows used to process data may incur costs. Additionally, storing data in the buckets after uploading can incur costs, too. For this vignette, we were able to upload the data for essentially nothing, while maintaining the workspace with the data costs $0.358 a day, according to our records in late February, 2026.

Check out the [Google Cloud Console](https://console.cloud.google.com/) for accurate reporting of your own costs.

You can also read more about [cloud costs](https://anvilproject.org/learn/control-cloud-costs/understanding-cloud-costs) on the AnVIL Portal.
:::

> ## Step One: Create your workspace

Reviewer 3: This is the only heading with "Step [number]" within this section

Author: Good catch - thanks!

```{r, echo=FALSE, fig.alt="You will need to give your new workspace a unique name and assign it to a billing project. The “anvil-outreach” billing project is used here as an example, but you don't have permission to use it. You’ll have to use one of your own. After filling out these two fields, click the “Quick Create Workspace” to create your workspace without enabling sharing or additional security options."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3c63398e684_33_0#slide=id.g3c63398e684_33_0")
```

You can read about Authorization Domains for workspace security in [this article](https://support.terra.bio/hc/en-us/articles/8527464803739-How-to-set-up-and-use-an-Authorization-Domain) in the Terra documentation.

Once you have created a workspace, AnVIL will take you to the workspace dashboard. 

## Open the Data Uploader

> When you switch to the Data tab in your Workspace, you'll see three headers on the left hand side: **TABLES**, **REFERENCE DATA**, and **WORKSPACE DATA**. At the moment all three of these sections are empty. We will use the Data Uploader tool to populate the **TABLES** section of the Workspace.

Reviewer 3: Might be worth adding a click to expand section with a brief description of what these three mean and why the tables section is the one getting populated with this vignette?

Reviewer 2: Agree that it would be useful to explain this or to link out to an explanation

Author: Done

Click on the box in the upper left hand corner that says "Import Data". From the dropdown menu, you'll want to choose "Open data uploader".

```{r, echo=FALSE, fig.alt="You can access the Data Uploader through the data tab on your Workspace. Data Uploader is one of the options given in the Import Data menu."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3c63398e684_33_3#slide=id.g3c63398e684_33_3")
```

When you first open the Data Uploader, you will be prompted to choose a workspace. Click on the workspace that you have just created.

```{r, echo=FALSE, fig.alt="The Data Uploader first asks you to select a workspace. Choose the one you just created."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3c63398e684_33_6#slide=id.g3c63398e684_33_6")
```

## Upload data and metadata

> After you have chosen your workspace, the Data Uploader then will then prompt you to create a data collection.  This is essentially a directory that contains all the samples described by a single metadata file. Click on "Create a New Collection".

Reviewer 3: ```suggestion
After you have chosen your workspace, the Data Uploader then will prompt you to create a data collection.  This is essentially a directory that contains all the samples described by a single metadata file. Click on "Create a New Collection".
```

```{r, echo=FALSE, fig.alt="In order to upload data, you will need to create a data collection. This is essentially a directory that contains all the samples described by a single metadata file. "}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3c63398e684_33_9#slide=id.g3c63398e684_33_9")
```

> You will need to choose a name for your data collection. In our example, we chose to name the collection "covid_sequences". Make sure to pick something descriptive enough so that you'll know what the samples are later! Here we have chosen to use covid_sequences, since we are uploading fasta files of covid spike protein sequences.

Reviewer 3: ```suggestion
You will need to choose a name for your data collection. In our example, we chose to name the collection "covid_sequences". Make sure to pick something descriptive enough so that you'll know what the samples are later! Here we have chosen to use "covid_sequences", since we are uploading FASTA files of COVID spike protein sequences.
```

```{r, echo=FALSE, fig.alt="Each collection should have a unique and descriptive name that easily describes the samples within it. Here we have chosen to use covid_sequences, since we are uploading fasta files of covid spike protein sequences."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3c63398e684_33_12#slide=id.g3c63398e684_33_12")
```

Next, click on the "Upload" button. The Data Uploader will prompt you to choose the files from your computer that you want to upload into your workspace bucket. Remember that every file must have a unique name!

```{r, echo=FALSE, fig.alt="Once you have created a new collection, you can upload your samples using the pop-up menu by clicking upload."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3c63398e684_33_18#slide=id.g3c63398e684_33_18")
```

> After you upload your data files, you can always remove files. In this example, we accidentally uploaded a fasta file called "sars_membrane_protein_omicron.fasta". To remove it, check the box next to the file name and click the "Delete" button.

Reviewer 3: ```suggestion
You can always remove files after uploading. In this example, we accidentally uploaded a FASTA file called "sars_membrane_protein_omicron.fasta". To remove it, check the box next to the file name and click the "Delete" button.
```

```{r, echo=FALSE, fig.alt="You can always remove files from your collection using the Data Uploader. We accidentally uploaded a file that we don’t need. In order to delete this file, check the box next to its name and choose Delete."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3c63398e684_33_21#slide=id.g3c63398e684_33_21")
```

Once you have verified all the files that you want in your workspace bucket have been uploaded, click the "Next" button so that you can load the sample metadata. 

> ```{r, echo=FALSE, fig.alt="Before you move on to uploading your metadata file, you should always verify that you have uploaded all the samples. Once you done this, choose Next."}

Reviewer 3: ```suggestion
```{r, echo=FALSE, fig.alt="Before you move on to uploading your metadata file, you should always verify that you have uploaded all the samples. Once you have done this, choose Next."}
```

ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3c63398e684_33_15#slide=id.g3c63398e684_33_15")
```

> Data Uploader requires you to drag and drop your metadata file onto a bar at the bottom of the page.

Reviewer 3: Does every data upload always need a metadata file? Assuming since the "entity:sample_id" column needs to have the file names this is probably a yes?

Author: Surprisingly no

Reviewer 3: Might be worth adding a note about that then? And whether you'd recommend it even though it's not required?

```{r, echo=FALSE, fig.alt="Data Uploader requires that you drag and drop your metadata file into a box at the bottom of the page."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3c63398e684_33_24#slide=id.g3c63398e684_33_24")
```

> Your metadata file should be saved as a tsv. The first line needs to contain the column names for the metadata. Additionally, the first column should be called "entity:sample_id". **Each sample ID should match the name of the files you uploaded earlier.** Remember to use tabs between each column.

Reviewer 3: Do the file extensions not need to be included?

Author: Nope!

> Your metadata file should be saved as a tsv. The first line needs to contain the column names for the metadata. Additionally, the first column should be called "entity:sample_id". **Each sample ID should match the name of the files you uploaded earlier.** Remember to use tabs between each column.

Reviewer 3: Is it worthwhile to add a sentence about recommendations on building the metadata file (e.g., use notepad or use excel/googlesheets/etc and save as a tsv?

> Your metadata file should be saved as a tsv. The first line needs to contain the column names for the metadata. Additionally, the first column should be called "entity:sample_id". **Each sample ID should match the name of the files you uploaded earlier.** Remember to use tabs between each column.

Reviewer 2: I'm not clear why "each sample ID should match the name of the files you uploaded earlier". The Data Uploader doesn't care, right? It just needs the IDs to be unique?

(I mean, obviously you should give them IDs that clearly indicate what they are, and the filenames work for that, I'm just wondering about the emphasis being made here.)

Author: Fair point! This was my misinterpretation of the documentation and has now been fixed.

```

The Data Uploader will display a table that displays your metadata. If it looks good, you can click "Create Table" to finish the file upload.

```{r, echo=FALSE, fig.alt="After dropping your metadata file into the Data Uploader, you will see a table that displays the metadata. If everything looks fine, click Create Table to finish the file upload."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3c63398e684_33_30#slide=id.g3c63398e684_33_30")
```

You can add multiple data collections to your workspace. If you go back to "Select A Collection", clicking on "Create New Collection" allows you to make an additional data collection. You will need to choose a unique name for your new collection. Our new data collection is called "covid_membrane".

```{r, echo=FALSE, fig.alt="You can add multiple data collections to a single workspace."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3c7f7fdd14a_0_10#slide=id.g3c7f7fdd14a_0_10")
```

> After you have created a new data collection, you can add your new files (and metadata). In this example, we loaded four additional fasta files that contain sequences of a CoVID membrane protein gene.

Reviewer 3: ```suggestion
After you have created a new data collection, you can add your new files (and metadata). In this example, we have loaded four additional FASTA files that contain sequences of a COVID membrane protein gene.
```

```{r, echo=FALSE, fig.alt="After you have created a new data collection, you can add your new files (and metadata). In this example, we loaded four additional fasta files that contain sequences of a CoVID membrane protein gene."}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3c7f7fdd14a_0_15#slide=id.g3c7f7fdd14a_0_15")
```

> 

Reviewer 3: Would it be worth adding an exercise here to create and add the metadata file for these sequences?

```

```{r, echo=FALSE, fig.alt='Image shows a screenshot of the workspace Data tab. The Files directory and link on the bottom left is highlighted. This image shows the collection called covid_membrane.'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3c7f7fdd14a_0_27#slide=id.g3c7f7fdd14a_0_27")
```

If you click on the name of the file, you can also see the details of the file.

```{r, echo=FALSE, fig.alt='Image shows a screenshot of the details of the files we successfully transferred.'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1lSUfsg_oja-Iqq5pTFD1VOuR32GLUq2l-sCBYNO3OTg/edit?slide=id.g3c7f7fdd14a_0_30#slide=id.g3c7f7fdd14a_0_30")
```

This is a good time to check that the sizes of the files you transferred match the sizes of the original files.

## Summary

- Create a workspace
- Open Data Uploader
- Upload data and metadata (this requires creating a data collection)
- Verify that files have been transferred

## Additional Resources

> Check out these AnVIL Portal articles for more details!

Reviewer 2: ```suggestion
Check out these articles for more details!
```

> - You can read more about the Data Uploader [here](https://support.terra.bio/hc/en-us/articles/4419428208411-Upload-data-and-populate-the-table-with-linked-file-paths). 

Reviewer 3: I'd maybe move this part to the "Open the Data Uploader" section.

> - You can read more about the Data Uploader [here](https://support.terra.bio/hc/en-us/articles/4419428208411-Upload-data-and-populate-the-table-with-linked-file-paths). 

Reviewer 3: ```suggestion
- You can read more [about the Data Uploader](https://support.terra.bio/hc/en-us/articles/4419428208411-Upload-data-and-populate-the-table-with-linked-file-paths). 
```

Reviewer 3: recommending for accessibliity

> - AnVIL datasets are organized using tables. [This post](https://support.terra.bio/hc/en-us/articles/360025758392-Managing-data-with-tables) explains the advantages of using tables with your data on AnVIL, as well as offers tips for customizing your data tables.

Reviewer 3: ```suggestion
- AnVIL datasets are organized using tables. [This post](https://support.terra.bio/hc/en-us/articles/360025758392-Managing-data-with-tables) explains the advantages of using tables with your data on AnVIL, as well as offers tips for customizing your data tables. **We especially recommend this resource because it demonstrates how to add Google bucket links to the metadata table after the files have been uploaded, enabling work with workflows down the road.**
```

> - If you have a lot of data, it's a good idea to estimate how much transfer time your need. Transfer a small file first and determine your transfer rate. Learn more about estimated transfer rates to Google Cloud via AnVIL [here](https://docs.cloud.google.com/architecture/migration-to-google-cloud-transferring-your-large-datasets#online_versus_offline_transfer). 

Reviewer 3: ```suggestion
- If you have a lot of data, it's a good idea to estimate how much transfer time you need. Transfer a small file first and determine your transfer rate. [Learn more about estimated transfer rates to Google Cloud via AnVIL](https://docs.cloud.google.com/architecture/migration-to-google-cloud-transferring-your-large-datasets#online_versus_offline_transfer). 
```
